### PR TITLE
Use Spring IoC to wire backend components

### PR DESCRIPTION
**Pull request by henryju**

### CODE REVIEW

**Comment by henryju** [comment]
@kirill-knize-sonarsource you might be interested to have a look

File: core/src/main/java/org/sonarsource/sonarlint/core/SonarLintBackendImpl.java [hunk 1]
```diff
@@ -45,79 +52,83 @@ public SonarLintBackendImpl(SonarLintClient client) {
   @Override
   public CompletableFuture<Void> initialize(InitializeParams params) {
     return CompletableFuture.runAsync(() -> {
-      if (initialized.compareAndSet(false, true)) {
-        sonarLintBackend = new InitializedSonarLintBackend(client, params);
+      if (initializeCalled.compareAndSet(false, true) && !initialized.get()) {
```
Comment (henryju) [inline core/src/main/java/org/sonarsource/sonarlint/core/SonarLintBackendImpl.java:55]: This change is not strictly related to the Spring migration, but is a way to make the initialization sequence safer. Let me know if you prefer to have it in a separate commit.

**Review by damien-urruty-sonarsource — APPROVED** [review]
LGTM. I find it a bit painful that for services we will have to use `Named`, `Singleton` and `Inject` every time, but I am not sure if this could be simplified.

**Comment by damien-urruty-sonarsource** [comment]
e.g. I don't know if we could have a single `@Service` annotation that would be sufficient

File: core/src/main/java/org/sonarsource/sonarlint/core/embedded/server/EmbeddedServer.java [hunk 2]
```diff
@@ -57,10 +65,19 @@ public class EmbeddedServer {
...
+  @Inject
   public EmbeddedServer(SonarLintClient client, ConnectionServiceImpl connectionService, AwaitingUserTokenFutureRepository awaitingUserTokenFutureRepository,
     ConfigurationServiceImpl configurationService, BindingSuggestionProviderImpl bindingSuggestionProvider, ServerApiProvider serverApiProvider,
-    TelemetryServiceImpl telemetryService, HostInfoDto clientInfo) {
+    TelemetryServiceImpl telemetryService, InitializeParams params) {
+    this(client, connectionService, awaitingUserTokenFutureRepository, configurationService, bindingSuggestionProvider, serverApiProvider, telemetryService, params.getHostInfo(),
+      params.shouldManageLocalServer());
+  }
+
+  public EmbeddedServer(SonarLintClient client, ConnectionServiceImpl connectionService, AwaitingUserTokenFutureRepository awaitingUserTokenFutureRepository,
```
[the remark this answers is not in the document]
Comment (henryju) [reply core/src/main/java/org/sonarsource/sonarlint/core/embedded/server/EmbeddedServer.java:78]: I wanted to avoid breaking changes in the unit tests (if any), but maybe it is not necessary.

**Comment by henryju** [comment]
@damien-urruty-sonarsource I tried to avoid Spring-specific annotations. `@Singleton` is not necessary here, because it is the default Spring behavior to create beans as singletons. But apparently this is not the standard of javax.inject, so I thought it would make it more explicit to see that all components are singletons.

`@Inject` is only required when there are multiple constructors.
